:bug: #2096 in tokenizer.decode, adds a space after special tokens for string format

mandubian opened the pull request:
This correction is cosmetic to correct the observed formatting issue.
No test was implemented because ideally composition of functions `encode.decode` should in theory return the original sentence. Yet there are some space strip (and lower-casing) in code so it's not certain to return exactly the original sentence with same spaces.

thomwolf commented on transformers/tokenization_utils.py:
```diff
@@ -1085,7 +1085,7 @@ def decode(self, token_ids, skip_special_tokens=False, clean_up_tokenization_spa
                     sub_texts.append(self.convert_tokens_to_string(current_sub_text))
                     current_sub_text = []
-                sub_texts.append(" " + token)
+                sub_texts.append(" " + token + " ")
```
Wouldn't this generate double spaces in the output string?

mandubian replied:
Actually, that's what I thought at start but no. Text tokens are joined with spaces in `convert_tokens_to_string` but then subtexts are joined with `''` and a space is missing after added tokens. Maybe my correction is too superficial and a rethink of this decoder formatting and join with spaces (or another separator) would be needed.

LysandreJik commented:
Hi @mandubian, thanks for opening a PR to fix this. I think this raises an issue when there are two new tokens added which are right after each other, as spaces get added before and after the tokens. Here's an example of the issue, with your change and based on #2096:

```py
from transformers import BertTokenizer
bert_tokenizer = BertTokenizer.from_pretrained('bert-base-cased')

bert_tokenizer.add_tokens(['[ENT]', '[TEN]'])
print(len(bert_tokenizer))

x = bert_tokenizer.encode("you are the [ENT] [TEN] with [ENT] and [ENT]")

print(bert_tokenizer.decode(x))
# outputs: [CLS] you are the [ENT]  [TEN] with [ENT] and [ENT] [SEP]
# with two spaces ----------------^^
```

mandubian commented:
You're right, I hadn't thought about the case of 2 consecutive tokens. Let's try to make it better.

mandubian commented:
@LysandreJik I've pushed a new version for discussion. Added tokens aren't prepended with space anymore but subtexts are. I've considered different solutions but none is perfect and a compromise has to be made. `convert_tokens_to_string` already joins with space between sub-strings but then those sub-strings and string corresponding to added_tokens need to be separated also by spaces.

if you have `[ABC] toto tata [DEF] [GHI]`, `decode.encode` returns the same string (except lower case).
But when you have less spaces `[ABC]toto tata [DEF][GHI]`, `decode.encode` returns `[ABC] toto tata [DEF] [GHI]` with more spaces.

I've added a test in `tokenizer_bert_test` but it's clearly not a very portable test, it depends on the tokenizer vocabulary.

LysandreJik commented:
This seems to work, thanks @mandubian! I pushed another commit on your fork to test every tokenizer.

mandubian commented:
@LysandreJik thks! I'll do that, do you prefer me to squash in one single commit too?

mandubian commented:
@LysandreJik (I've deleted my previous message from tonight, I had misread your message on mobile :D) Just to know, don't you squash commits in general?
